matlab/src/cpp/arrow/matlab/array/proxy/list_array.cc

```diff
@@ -0,0 +1,103 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/matlab/array/proxy/list_array.h"
+#include "arrow/matlab/array/proxy/numeric_array.h"
+#include "arrow/matlab/array/proxy/wrap.h"
+#include "arrow/matlab/error/error.h"
+#include "libmexclass/proxy/ProxyManager.h"
+
+namespace arrow::matlab::array::proxy {
+
+    ListArray::ListArray(std::shared_ptr<arrow::ListArray> list_array) : proxy::Array{std::move(list_array)} {
+        REGISTER_METHOD(ListArray, getValues);
+        REGISTER_METHOD(ListArray, getOffsets);
+    }
+
+    libmexclass::proxy::MakeResult ListArray::make(const libmexclass::proxy::FunctionArguments& constructor_arguments) {
+        namespace mda = ::matlab::data;
+        using libmexclass::proxy::ProxyManager;
+        using Int32ArrayProxy = ::arrow::matlab::array::proxy::NumericArray<arrow::Int32Type>;
+        using ListArrayProxy = ::arrow::matlab::array::proxy::ListArray;
+        using ArrayProxy = ::arrow::matlab::array::proxy::Array;
+
+        mda::StructArray opts = constructor_arguments[0];
+        const mda::TypedArray<uint64_t> offsets_proxy_id_mda = opts[0]["OffsetsProxyID"];
+        const mda::TypedArray<uint64_t> values_proxy_id_mda = opts[0]["ValuesProxyID"];
+        const mda::TypedArray<bool> validity_bitmap_mda = opts[0]["Valid"];
+
+        const auto offsets_proxy_id = offsets_proxy_id_mda[0];
+        const auto values_proxy_id = values_proxy_id_mda[0];
+
+        const auto offsets_proxy = std::static_pointer_cast<Int32ArrayProxy>(ProxyManager::getProxy(offsets_proxy_id));
+        const auto values_proxy = std::static_pointer_cast<ArrayProxy>(ProxyManager::getProxy(values_proxy_id));
+
+        const auto offsets = offsets_proxy->unwrap();
+        const auto values = values_proxy->unwrap();
+
+        // Pack the validity bitmap values.
+        MATLAB_ASSIGN_OR_ERROR(auto validity_bitmap_buffer,
+                               bit::packValid(validity_bitmap_mda),
+                               error::BITPACK_VALIDITY_BITMAP_ERROR_ID);
+
+        // Create a ListArray from values and offsets.
+        MATLAB_ASSIGN_OR_ERROR(auto array,
```

Comment:
Maybe not in this PR, but should we add an option to validate the `Offsets` and `Values` array are correct?

Reply:
Ah, yes! Thanks for the reminder! I meant to create a follow up issue for this, but I forgot.

I've created #38361 to track this.

Reply:
Marking this as resolved.